caas/kubernetes/provider/k8s.go

```diff
@@ -503,6 +511,50 @@ func (k *kubernetesClient) DeleteService(appName string) (err error) {
 	return errors.Trace(k.deleteDeployment(appName))
 }
 
+func (k *kubernetesClient) EnsureCrd(appName string, podSpec *caas.PodSpec) error {
+	for _, t := range podSpec.CustomResourceDefinitions {
+		crd, err := k.ensureCrdTemplate(&t)
+		if err != nil {
+			return errors.Annotate(err, "ensuring custom resource definition template")
+		}
+		logger.Debugf("created crd %q", crd.ObjectMeta.Name)
+	}
+	return nil
+}
+
+func (k *kubernetesClient) ensureCrdTemplate(t *caas.CustomResourceDefinition) (crd *apiextensionsv1beta1.CustomResourceDefinition, err error) {
+	singularName := strings.ToLower(t.Kind)
+	pluralName := fmt.Sprintf("%ss", singularName)
+	crdIn := &apiextensionsv1beta1.CustomResourceDefinition{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      fmt.Sprintf("%s.%s", pluralName, t.Group),
+			Namespace: k.namespace,
+		},
+		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
+			Group:   t.Group,
+			Version: t.Version,
+			Scope:   apiextensionsv1beta1.ResourceScope(t.Scope),
+			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+				Plural:   pluralName,
+				Kind:     t.Kind,
+				Singular: singularName,
+			},
+			Validation: &apiextensionsv1beta1.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextensionsv1beta1.JSONSchemaProps{
+					Properties: t.Validation.Properties,
+				},
+			},
+		},
+	}
+	apiextensionsV1beta1 := k.apiextensionsClient.ApiextensionsV1beta1()
+	crd, err = apiextensionsV1beta1.CustomResourceDefinitions().Create(crdIn)
+	if k8serrors.IsAlreadyExists(err) {
```

Comment:
In other places we do this the other way around - attempt to Update() and if we get a k8sNotFound error, we then create.